public/externalLibs/visualizer/visualizer.js

```diff
@@ -709,6 +719,11 @@
    *  Then shift it to the left end.
    */
   function draw(xs) {
+    stage = new Kinetic.Stage({
+      width: 1000,
```

Comment:
Can I ask why width defaults to 1000? This creates some padding on the right that shouldnt be there.

![image](https://user-images.githubusercontent.com/17170991/53719572-78da4680-3e99-11e9-9eda-079d43555289.png)

Reply:
I'm not really sure how to go about finding the width. It seems like it would be tricky for lists that contains pointers to other lists...

Reply:
I think you can find the width of the required window by visiting the lowest node on the right fringe of the tree. You start at the root and go right at each step until you reach a leaf. I think with the current tree drawing algorithm, you will find that that leaf is the node with the rightmost position.

Reply:
The problem is drawing like [these](https://sourceacademy.nus.edu.sg/playground#chap=4&ext=NONE&prgrm=MYewdgzgLgBAhjAvDANgS2gCnVgjAGhgCZCcpMBmASlI3IBZCBWGmABzjQCdMA2QgOxVWJGNQDcAKFCRYAIySo6mOIVUxcABkK4C8KlIAmXOAHcA%2BmUxyDQA), where there are both left and right nodes to account for, as well as multiple references to the same object. Nonetheless I think I may have a solution, which I'm working on now.

Reply:
Ok, I've sorted it out! Both width and height are now calculated dynamically.

Reply:
> The problem is drawings like [these](https://sourceacademy.nus.edu.sg/playground#chap=4&ext=NONE&prgrm=MYewdgzgLgBAhjAvDANgS2gCnVgjAGhgCZCcpMBmASlI3IBZCBWGmABzjQCdMA2QgOxVWJGNQDcAKFCRYAIySo6mOIVUxcABkK4C8KlIAmXOAHcA%2BmUxyDQA), where there are both left and right nodes to account for, as well as multiple references to the same object. Nonetheless I think I may have a solution, which I'm working on now.

Wow, that's a weird example. Great that you worked it out.